CIPs/0008.md

```diff
@@ -64,14 +63,56 @@ I considered doing storing all data on IPFS, however the tooling around IPFS as
+
+Signing over the path as well as the message itself prevents messages from being shuffled around by an attacker; a message is only valid in the intended context. The path should be normalized to the UTF8 encoding of the absolute URL path component, as specified in RFC 3986 § 3.3. Note that path does not include the storage root, query parameters, or fragment.
+
+It is worth noting that this does not protect against the adversarial storage provider withholding or returning outdated data. Additionally, this does not protect against the adversary swapping out files. If the order, including the existence, of the data is critical, then the developer should consider falling back to on-chain data, as consensus on the current view of the data is required. A possible partial workaround is to add expiry to the schema.
+
+### Privacy
+
+We commonly expect users to share data only with a set of authorized peers. A profile picture, under the path /profile/picture might be shared only with the user’s friends. Creating and publishing a number of ciphertexts to a related path (e.g. /profile/picture/{recipient address}), each encrypted with a friend's public key would be one way to achieve this. It would ensure data integrity and confidentiality.
+
+A problem with indexing the ciphertexts by recipient is that it reveals publicly every recipient of a given message. In the case of profile pictures, this would make public every “friend” with whom the picture is shared, and in the case of payment requests it would reveal that a request is pending from the user to a given peer. This data may be valuable to attackers looking to craft a spear-phishing attack, and may violate some users’ expectations of privacy.
+
+To solve this we present a scheme that maintains both sender and receiver privacy, to compute the storage path it requires the publisher's private data encryption key and the receivers public data encryption key. The path can be computed as follows:
+
+```
+label = PRF(ECDH(A, B), A || B || data_path)
+ciphertext path = "/ciphertexts/" || hex(label)
+```
+
+Where A is the sender’s public key, B is the recipient’s public key, ECDH is a non-interactive key exchange requiring the private key for either A or B to evaluate, PRF is the pseudorandom function HMAC-BLAKE2s-256, "ciphertexts" is a cosmetic path URL path prefix, and data pathis the URL path suffix. ECDH produces a secp256k1 elliptic curve point, which is encoded and used as the 64B key for PRF. Elliptic curve points (e.g. public keys and shared secrets) are represented in SEC1-v2 compressed representation. The message input to the PRF consists of the concatenation of the public key of the sender and the public key of the recipient (both in SEC1-v2 compressed representation), and the UTF8 encoding of the data path.
```

Comment:
If I understand our implementation correctly, the line about using the ECDH output as a 64B key is slightly inaccurate. I believe we get back from the node crypto API the 33B compressed form, and should make that the standard. https://github.com/celo-org/celo-monorepo/pull/5091/files#diff-baa5b1d1de4d5708ed55809c3c630d36442ddf5fb84eb26e62ef9f52a0930c3eR10

Note: This does not represent a reduction in entropy.
```suggestion
Where A is the sender’s public key, B is the recipient’s public key, ECDH is a non-interactive key exchange requiring the private key for either A or B to evaluate, PRF is the pseudorandom function HMAC-BLAKE2s-256, "ciphertexts" is a cosmetic path URL path prefix, and data pathis the URL path suffix. ECDH produces a secp256k1 elliptic curve point, which is encoded  in SEC1-v2 compressed format and used as the key for PRF. Elliptic curve points (e.g. public keys and shared secrets) are represented in SEC1-v2 compressed representation. The message input to the PRF consists of the concatenation of the public key of the sender and the public key of the recipient (both in SEC1-v2 compressed representation), and the UTF8 encoding of the data path.
```